packages/compartment-mapper/src/zip/buffer-reader.js

```diff
@@ -0,0 +1,225 @@
+const bitsTooHigh = (-1 << (53 - 32)) >> 0;
+
+export class BufferReader {
+  /**
+   * @param {ArrayBuffer} data
+   */
+  constructor(data) {
+    this.data = new Uint8Array(data);
+    this.length = this.data.length;
+    this.index = 0;
+    this.offset = 0;
+  }
+
+  /**
+   * @param {number} index
+   * @return {boolean} whether the read head can move to the given absolute
+   * index.
+   */
+  canSeek(index) {
+    return index >= 0 && this.offset + index <= this.length;
+  }
+
+  /**
+   * @param {number} index the index to check.
+   * @throws {Error} an Error if the index is out of bounds.
+   */
+  assertCanSeek(index) {
+    if (!this.canSeek(index)) {
+      throw new Error(
+        `End of data reached (data length = ${this.length}, asked index ${index}`
+      );
+    }
+  }
+
+  /**
+   * @param {number} index
+   * @return {number} prior index
+   */
+  seek(index) {
+    const restore = this.index;
+    this.assertCanSeek(index);
+    this.index = index;
+    return restore;
+  }
+
+  /**
+   * @param {number} size
+   * @returns {Uint8Array}
+   */
+  peek(size) {
+    // Clamp size.
+    size = Math.min(this.length - this.index, size);
+    if (size === 0) {
+      // in IE10, when using subarray(idx, idx), we get the array [0x00] instead of [].
+      return new Uint8Array(0);
+    }
+    const result = this.data.subarray(
+      this.offset + this.index,
+      this.offset + this.index + size
+    );
+    return result;
+  }
+
+  /**
+   * @param {number} offset
+   */
+  canRead(offset) {
+    return this.canSeek(this.index + offset);
+  }
+
+  /**
+   * Check that the offset will not go too far.
+   * @param {number} offset the additional offset to check.
+   * @throws {Error} an Error if the offset is out of bounds.
+   */
+  assertCanRead(offset) {
+    this.assertCanSeek(this.index + offset);
+  }
+
+  /**
+   * Get raw data without conversion, <size> bytes.
+   * @param {number} size the number of bytes to read.
+   * @return {Uint8Array} the raw data.
+   */
+  read(size) {
+    this.assertCanRead(size);
+    const result = this.peek(size);
+    this.index += size;
+    return result;
+  }
+
+  /**
+   * @returns {number}
+   */
+  readUint8() {
+    this.assertCanRead(1);
+    const value = this.data[this.offset + this.index];
+    this.index += 1;
+    return value;
+  }
+
+  /**
+   * @returns {number}
+   */
+  readUint16LE() {
+    this.assertCanRead(2);
+    const index = this.offset + this.index;
+    const a = this.data[index + 0];
+    const b = this.data[index + 1];
+    const value = (b << 8) | a;
+    this.index += 2;
+    return value;
+  }
+
+  /**
+   * @returns {number}
+   */
+  readUint32LE() {
+    this.assertCanRead(4);
+    const index = this.offset + this.index;
+    const a = this.data[index + 0];
+    const b = this.data[index + 1];
+    const c = this.data[index + 2];
+    const d = this.data[index + 3];
+    const value = ((d << 24) >>> 0) + ((c << 16) | (b << 8) | a);
```

Comment:
Why is the high-order byte treated specially? Is it because the high-order bit might convince `<< 24` that this is a negative number?

Reply:
Ah, these are necessary for any semblance of Zip64 support, but I will move this to my scratch for later. There’s a better way to do this with a scratch BitUint64Array, but you’re right that the `>>>` does ensure the result is unsigned.